File: AkkaWithZeroMQ/src/main/scala/org/akka/essentials/zeromq/example4/PushActor.scala
package org.akka.essentials.zeromq.example4
import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorLogging
import akka.actor.Cancellable
import akka.util.duration.intToDurationInt
import akka.zeromq.Bind
import akka.zeromq.Frame
import akka.zeromq.Listener
import akka.zeromq.SocketType
import akka.zeromq.ZMQMessage
import akka.zeromq.ZeroMQExtension

case class Tick

class PushActor extends Actor with ActorLogging {
	val pushSocket = ZeroMQExtension(context.system).newSocket(SocketType.Push, Bind("tcp://127.0.0.1:1234"), Listener(self))

	var count = 0
	var cancellable: Cancellable = null
	override def preStart() {
		cancellable = context.system.scheduler.schedule(1 second, 1 second, self, Tick)
	}

	def receive: Receive = {
		case Tick =>
			count += 1
			var payload = "Hi there! (" + count + ")"
			pushSocket ! ZMQMessage(Seq(Frame(payload)))
			if (count == 5) {
				cancellable.cancel()
			}
	}
}

File: AkkaWithZeroMQ/src/main/java/org/akka/essentials/zeromq/example4/PushActor.java
package org.akka.essentials.zeromq.example4;

import akka.actor.ActorRef;
import akka.actor.Cancellable;
import akka.actor.UntypedActor;
import akka.util.Duration;
import akka.zeromq.Bind;
import akka.zeromq.Frame;
import akka.zeromq.Listener;
import akka.zeromq.SocketOption;
import akka.zeromq.ZMQMessage;
import akka.zeromq.ZeroMQExtension;

public class PushActor extends UntypedActor {
	public static final Object TICK = "TICK";
	int count = 0;
	Cancellable cancellable;
	ActorRef pushSocket = ZeroMQExtension.get(getContext().system())
			.newPushSocket(
					new SocketOption[] { new Bind("tcp://127.0.0.1:1237"),
							new Listener(getSelf()) });

	@Override
	public void preStart() {

		cancellable = getContext()
				.system()
				.scheduler()
				.schedule(Duration.parse("1 second"),
						Duration.parse("1 second"), getSelf(), TICK);
	}

	@Override
	public void onReceive(Object message) throws Exception {
		if (message.equals(TICK)) {

			count++;
			pushSocket.tell(new ZMQMessage(
					new Frame("Hi there (" + count + ")")));
			if (count == 5)
				cancellable.cancel();
		}

	}
}
